### ISSUE TITLE
[DOXIASITETOOLS-301] Automatically remove the 0-byte site descriptors from the local repo

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-doxia-sitetools#551]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** opened **[DOXIASITETOOLS-301](https://issues.apache.org/jira/browse/DOXIASITETOOLS-301?redirect=false)** and commented

As older versions created 0-byte site descriptors in the local repo those should be automatically removed if detected. Otherwise you run into the following error

```
[ERROR] Failed to execute goal
org.apache.maven.plugins:maven-site-plugin:4.0.0-M6:site
(generate-site) on project openmeetings-parent: SiteToolException:
Error reading site descriptor: input contained no data -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to
execute goal org.apache.maven.plugins:maven-site-plugin:4.0.0-M6:site
(generate-site) on project openmeetings-parent: SiteToolException:
Error reading site descriptor
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2
(MojoExecutor.java:347)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute
(MojoExecutor.java:330)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:213)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:175)
   at org.apache.maven.lifecycle.internal.MojoExecutor.access$000
(MojoExecutor.java:76)
   at org.apache.maven.lifecycle.internal.MojoExecutor$1.run
(MojoExecutor.java:163)
   at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute
(DefaultMojosExecutionStrategy.java:39)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:160)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:105)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:73)
   at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build
(SingleThreadedBuilder.java:53)
   at org.apache.maven.lifecycle.internal.LifecycleStarter.execute
(LifecycleStarter.java:118)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
   at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
   at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
   at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
   at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke
(NativeMethodAccessorImpl.java:77)
   at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke
(DelegatingMethodAccessorImpl.java:43)
   at java.lang.reflect.Method.invoke (Method.java:568)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced
(Launcher.java:282)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launch
(Launcher.java:225)
   at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode
(Launcher.java:406)
   at org.codehaus.plexus.classworlds.launcher.Launcher.main
(Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException:
SiteToolException: Error reading site descriptor
   at org.apache.maven.plugins.site.descriptor.AbstractSiteDescriptorMojo.prepareDecorationModel
(AbstractSiteDescriptorMojo.java:94)
   at org.apache.maven.plugins.site.render.AbstractSiteRenderingMojo.createSiteRenderingContext
(AbstractSiteRenderingMojo.java:246)
   at org.apache.maven.plugins.site.render.SiteMojo.renderLocale
(SiteMojo.java:138)
   at org.apache.maven.plugins.site.render.SiteMojo.execute (SiteMojo.java:123)
   at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo
(DefaultBuildPluginManager.java:126)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2
(MojoExecutor.java:342)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute
(MojoExecutor.java:330)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:213)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:175)
   at org.apache.maven.lifecycle.internal.MojoExecutor.access$000
(MojoExecutor.java:76)
   at org.apache.maven.lifecycle.internal.MojoExecutor$1.run
(MojoExecutor.java:163)
   at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute
(DefaultMojosExecutionStrategy.java:39)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:160)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:105)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:73)
   at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build
(SingleThreadedBuilder.java:53)
   at org.apache.maven.lifecycle.internal.LifecycleStarter.execute
(LifecycleStarter.java:118)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
   at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
   at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
   at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
   at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke
(NativeMethodAccessorImpl.java:77)
   at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke
(DelegatingMethodAccessorImpl.java:43)
   at java.lang.reflect.Method.invoke (Method.java:568)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced
(Launcher.java:282)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launch
(Launcher.java:225)
   at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode
(Launcher.java:406)
   at org.codehaus.plexus.classworlds.launcher.Launcher.main
(Launcher.java:347)
Caused by: org.apache.maven.doxia.tools.SiteToolException: Error
reading site descriptor
   at org.apache.maven.doxia.tools.DefaultSiteTool.readDecorationModel
(DefaultSiteTool.java:1026)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:947)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:979)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:386)
   at org.apache.maven.plugins.site.descriptor.AbstractSiteDescriptorMojo.prepareDecorationModel
(AbstractSiteDescriptorMojo.java:91)
   at org.apache.maven.plugins.site.render.AbstractSiteRenderingMojo.createSiteRenderingContext
(AbstractSiteRenderingMojo.java:246)
   at org.apache.maven.plugins.site.render.SiteMojo.renderLocale
(SiteMojo.java:138)
   at org.apache.maven.plugins.site.render.SiteMojo.execute (SiteMojo.java:123)
   at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo
(DefaultBuildPluginManager.java:126)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2
(MojoExecutor.java:342)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute
(MojoExecutor.java:330)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:213)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:175)
   at org.apache.maven.lifecycle.internal.MojoExecutor.access$000
(MojoExecutor.java:76)
   at org.apache.maven.lifecycle.internal.MojoExecutor$1.run
(MojoExecutor.java:163)
   at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute
(DefaultMojosExecutionStrategy.java:39)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:160)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:105)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:73)
   at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build
(SingleThreadedBuilder.java:53)
   at org.apache.maven.lifecycle.internal.LifecycleStarter.execute
(LifecycleStarter.java:118)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
   at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
   at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
   at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
   at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke
(NativeMethodAccessorImpl.java:77)
   at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke
(DelegatingMethodAccessorImpl.java:43)
   at java.lang.reflect.Method.invoke (Method.java:568)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced
(Launcher.java:282)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launch
(Launcher.java:225)
   at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode
(Launcher.java:406)
   at org.codehaus.plexus.classworlds.launcher.Launcher.main
(Launcher.java:347)
Caused by: java.io.EOFException: input contained no data
   at org.codehaus.plexus.util.xml.pull.MXParser.fillBuf (MXParser.java:3898)
   at org.codehaus.plexus.util.xml.pull.MXParser.more (MXParser.java:3980)
   at org.codehaus.plexus.util.xml.pull.MXParser.parseProlog
(MXParser.java:1774)
   at org.codehaus.plexus.util.xml.pull.MXParser.nextImpl (MXParser.java:1757)
   at org.codehaus.plexus.util.xml.pull.MXParser.next (MXParser.java:1375)
   at org.apache.maven.doxia.site.decoration.io.xpp3.DecorationXpp3Reader.read
(DecorationXpp3Reader.java:1415)
   at org.apache.maven.doxia.site.decoration.io.xpp3.DecorationXpp3Reader.read
(DecorationXpp3Reader.java:579)
   at org.apache.maven.doxia.site.decoration.io.xpp3.DecorationXpp3Reader.read
(DecorationXpp3Reader.java:594)
   at org.apache.maven.doxia.tools.DefaultSiteTool.readDecorationModel
(DefaultSiteTool.java:1022)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:947)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:979)
   at org.apache.maven.doxia.tools.DefaultSiteTool.getDecorationModel
(DefaultSiteTool.java:386)
   at org.apache.maven.plugins.site.descriptor.AbstractSiteDescriptorMojo.prepareDecorationModel
(AbstractSiteDescriptorMojo.java:91)
   at org.apache.maven.plugins.site.render.AbstractSiteRenderingMojo.createSiteRenderingContext
(AbstractSiteRenderingMojo.java:246)
   at org.apache.maven.plugins.site.render.SiteMojo.renderLocale
(SiteMojo.java:138)
   at org.apache.maven.plugins.site.render.SiteMojo.execute (SiteMojo.java:123)
   at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo
(DefaultBuildPluginManager.java:126)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2
(MojoExecutor.java:342)
   at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute
(MojoExecutor.java:330)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:213)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:175)
   at org.apache.maven.lifecycle.internal.MojoExecutor.access$000
(MojoExecutor.java:76)
   at org.apache.maven.lifecycle.internal.MojoExecutor$1.run
(MojoExecutor.java:163)
   at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute
(DefaultMojosExecutionStrategy.java:39)
   at org.apache.maven.lifecycle.internal.MojoExecutor.execute
(MojoExecutor.java:160)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:105)
   at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject
(LifecycleModuleBuilder.java:73)
   at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build
(SingleThreadedBuilder.java:53)
   at org.apache.maven.lifecycle.internal.LifecycleStarter.execute
(LifecycleStarter.java:118)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:260)
   at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:172)
   at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:100)
   at org.apache.maven.cli.MavenCli.execute (MavenCli.java:821)
   at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:270)
   at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
   at jdk.internal.reflect.NativeMethodAccessorImpl.invoke
(NativeMethodAccessorImpl.java:77)
   at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke
(DelegatingMethodAccessorImpl.java:43)
   at java.lang.reflect.Method.invoke (Method.java:568)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced
(Launcher.java:282)
   at org.codehaus.plexus.classworlds.launcher.Launcher.launch
(Launcher.java:225)
   at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode
(Launcher.java:406)
   at org.codehaus.plexus.classworlds.launcher.Launcher.main
(Launcher.java:347)
```

This has been reported in https://the-asf.slack.com/archives/C7Q9JB404/p1680156986876729?thread_ts=1680110297.233759&cid=C7Q9JB404 and in https://lists.apache.org/thread/8w6c5l2pz2hbv918os0f8w58l4jnflyw

---

**Affects:** 2.0.0-M6

**Issue Links:**
- [DOXIASITETOOLS-305](https://issues.apache.org/jira/browse/DOXIASITETOOLS-305) Removal of 0-byte site descriptors from the local repo does not put remote repos into consideration
 (_**"causes"**_)
- [DOXIASITETOOLS-294](https://issues.apache.org/jira/browse/DOXIASITETOOLS-294) Replace legacy artifact resolution with Maven Resolver

- [DOXIASITETOOLS-294](https://issues.apache.org/jira/browse/DOXIASITETOOLS-294) Replace legacy artifact resolution with Maven Resolver
 (_**"is broken by"**_)

**Remote Links:**
- [GitHub Pull Request #104
](https://github.com/apache/maven-doxia-sitetools/pull/104)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please provide the command you have run and the verbose output of the build and I will have a look. I will refine docs later this day.

**Comment by jira-importer** [comment]
**[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson)** commented

I can't provide the output. The command was just mvn site.

I think I can see from the output what is going on. There is a parent project, which consists only of a pom.  It was built and installed with (among other plugins) maven-site-plugin version 4.0.0-M6.  In my local repository, I see .../project-name-20-site.xml.  It has no bytes in it.

The project I'm building inherits from the parent.  I see during mvn site:

`[DEBUG] Looking for site descriptor of level 1 parent project: com.foo:project-name:pom:20`
`[DEBUG] Reading parent level 1 site descriptor from /Users/ljnelson/.m2/repository/com/foo/project-name/20/project-name-20-site.xml`

The build fails with SiteToolException: Error reading site descriptor: input contained no data.

As noted above, this file has no bytes in it so my project build fails.

Perhaps when you said "all site.xml files" you meant "all files in your .m2 repository that end with 'site.xml'"?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Hence, the command refinement in the linked documentation.

**Comment by jira-importer** [comment]
**[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson)** commented

Thank you. As noted above, the parent was created with maven-site-plugin version 4.0.0-M6, so it is not, to my knowledge, an "older version" as described in the documentation. Would you agree?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

In this case, the version does not matter at the moment because it is a runtime issue, not a deploy time one. The model itself is still unchanged. I hope this answers your question.

**Comment by jira-importer** [comment]
**[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson)** commented

I see. So if some process ever places a zero-byte site descriptor anywhere in my `~/.m2/repository` (in my project graph) for any reason, subsequent usage of (for example) the maven-site-plugin version 4.0.0-M6 will fail? I am not trying to criticize, just trying to understand the scope of the problem.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Note: It is not on the project graph, unrelated. This is a separate resolution.

Now, let me elaborate how Maven Resolver works here since it is just an artifact (site descriptor): If an artifact is not in the local repo, read no file exists, regardless of the size, go out and try to download, persist to local disk and check the checksum. If all fine, report if not, fail with an exception and mark as given out for resolution, will not try next time unless (a) you explicitly request it or (b) the wait time is over (see [update policy](https://maven.apache.org/pom.html#Repositories)). This is a great process not to reinvent the wheel.

Here is the old approach which combines reprecated Maven Artifact API and a band aid: Try to resolve, not found? Abuse the local site descriptor file as a zero byte marker which will force the artifact API not to try to resolve again because the file is there since the assumption is, if it wasn't there before, it won't appear at all. Might be true for releases, but not for snapshots. Using an artifact file as as marker is not a good solution.

Now pair both: Resolver sees the zero byte pseudo marker file and reports: Yep, here is your artifact. XML readers an empty file, there is your exception.

Remedy: purge once and use proper Maven Resolver approach.

**Comment by jira-importer** [comment]
**[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson)** commented

Thank you; I see now. Unfortunately as a previous commenter noted above I cannot ensure that all builds of everything will always use the Maven Resolver approach, so those zero-byte site descriptors may keep showing up. Is the remedy nevertheless still to repeatedly destroy my local repository, over and over again? If so, I guess I'll just roll back to the 3.12.1 version of the maven-site-plugin to avoid this.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Yes, correct. That is the cosequence for now. The issue from the past bites us now. An all or nothing approach. How big is the impact for you for a mixed project setup in the same local repo? Are you willing to try a patched version for a cross-major version compat?

**Comment by jira-importer** [comment]
**[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson)** commented

Sure, I am willing. I think that I may be only one user at the moment with this problem, but there are going to be **lots** of others, since my use case is not particularly complicated, so anything to prevent users from having to repeatedly destroy their local repositories is a good thing.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Laird Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ljnelson), please build from source: https://github.com/apache/maven-doxia-sitetools/pull/104, install into your local repo and point Maven Site Plugin to use this snapshot. When running in verbose mode you should see the debug message. It works for me on `maven-site` repo.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [60e5b6748d616cbbbf4f486ccf09da4c80b016a9](https://gitbox.apache.org/repos/asf?p=maven-doxia-sitetools.git;a=commit;h=60e5b6748d616cbbbf4f486ccf09da4c80b016a9).
